### DISCUSSION TITLE
wire:key not working for containers moving between different parent containers (which was working in v2)

### DISCUSSION THREAD

**Discussion by plorenz-etes** [livewire/livewire#7518] [7. Bugs (archived)]
### Livewire version

v3.2.6

### Laravel version

v10.35.0

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

1. Clone the example repository: `git clone https://github.com/plorenz-etes/livewire-wire-key-example.git`
2. Install composer packages: `composer install`
3. Create config file: `cp .env.example .env`
4. Generate application key: `php artisan key:generate`
5. Run `php artisan serve`
6. Visit http://127.0.0.1:8000 (or the corresponding link working for you)
7. Click the button `Flip $wrapElements`

### Repository or Wirebox URL

https://github.com/plorenz-etes/livewire-wire-key-example

### Additional information

The example outputs a list with `@foreach` containing user objects. All objects also receive a uniqid in the template.
Because the uniqid is contained inside a `wire:ignore` div container, the uniqid should never change as long as the DOM element with the corresponding `wire:key` exists somewhere in the Livewire component.

However, if the whole `@foreach` is wrapped inside another `div` container when re-rendering the component, all of the div containers with `wire:key` will be completely re-created which results in a changing random ID once you click the button.

This was working differently in Livewire v2 which detected that the containers with `wire:key` had been moved and just moved them around in the DOM instead of re-creating the elements.

If this is intended behavior, it should be mentioned in the docs (and the upgrade guide). This is a breaking change for a larger project and I haven't found a solution to resolve this problem yet.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by joshhanley** [comment]
@plorenz-etes yeah that is correct, in V2, we used morphdom under the hood, which treated `wire:key` as unique across the whole DOM just like IDs need to be unique.

In V3, one of the breaking changes that we now use Alpine morph under the hood, where `wire:key` are local to that parent element.

So this is probably why you are experiencing this behaviour.

PRs to the upgrade guide to add details around this would be welcome! 🙂

**Reply by plorenz-etes** [reply]
@joshhanley Thank you your getting back to me. Is there any convenient way to compensate this new behavior?

The project I'm working on is a video conference tool which consists of a Livewire component containing multiple video players (one for each participant). When a participant starts sharing their screen, all player windows will be sorted into two different div containers (the first being much bigger to view the shared screen and the other one much smaller for the rest of the video players).

This mechanism doesn't work anymore now because instead of moving the `<video>` elements they will be re-created. This will drop the connection to the stream as well as all settings made to the video player (e.g. volume controls, etc.).

**Reply by joshhanley** [reply]
@plorenz-etes hmm, one way I could imagine doing it, is to actually make the changes on the front end, moving the video elements around using javascript, and then making a request to Livewire, so it re-renders the page with the video elements in the same place. Which should work in theory, as Morph will see the elements in the same place (back end re-render matching with current javascript render), so not actually touch them.

The other way you could approach it is to use CSS Grid and so all the video elements are within a single element, allowing you to arrange and position them without ever actually moving them outside the parent element.

**Reply by plorenz-etes** [reply]
@joshhanley Thank you for your input! Regarding the first idea I'm a bit worried that two participants could join almost at the same time leading to a Livewire update while the JS part is moving around elements, but I'll see what I can do!

**Reply by joshhanley** [reply]
@plorenz-etes no worries! 🙂 Yeah that's something you'll have to find a way through. Maybe Livewire isn't the best choice for that set up anyway, so using javascript would be a better option (either Alpine or plain JS, or even another front end framework), and use `wire:ignore` to ensure Livewire doesn't mess with it.
